Cython/Utility/ImportExport.c

```diff
@@ -871,15 +871,31 @@ static int __Pyx_MergeVtables(PyTypeObject *type) {
     return -1;
 }
 
+/////////////// ImportNumPyArray.module_state_decls //////////////
+//@requires: MemoryView_C.c::Atomics
 
-/////////////// ImportNumPyArray.proto ///////////////
+#if CYTHON_COMPILING_IN_CPYTHON_FREETHREADING && CYTHON_ATOMICS
+__pyx_atomic_ptr_type __pyx_numpy_ndarray;
+#else
+// If freethreading but not atomics, then this is guarded by ndarray_mutex
+// in __Pyx__ImportNumPyArrayTypeIfAvailable
+PyObject *__pyx_numpy_ndarray;
+#endif
 
-static PyObject *__pyx_numpy_ndarray = NULL;
+/////////////// ImportNumPyArray.proto ///////////////
 
 static PyObject* __Pyx_ImportNumPyArrayTypeIfAvailable(void); /*proto*/
 
 /////////////// ImportNumPyArray.cleanup ///////////////
-Py_CLEAR(__pyx_numpy_ndarray);
+
+// I'm not actually worried about thread-safety in the cleanup function.
+// The CYTHON_ATOMICS code is only for the type-casting.
+#if CYTHON_COMPILING_IN_CPYTHON_FREETHREADING && CYTHON_ATOMICS
+Py_XDECREF((PyObject*)__pyx_atomic_pointer_load_acquire(&CGLOBAL(__pyx_numpy_ndarray)));
+__pyx_atomic_pointer_exchange(&CGLOBAL(__pyx_numpy_ndarray), NULL);
```

Comment:
If possible, I'd still use a "get-zero-decref" procedure here like `Py_CLEAR()` does, just to be on the safe side. Is there an atomic swap command? Like `local = NULL; atomic_swap(global, local); Py_XDECREF(local);`.